File: cns/exp3.cpp
#include <iostream>
#include <string>
#include <unordered_map>
#include <algorithm>

using namespace std;

string caesar_decrypt(const string& ciphertext, int shift) {
    string decrypted_text = "";
    for (char c : ciphertext) {
        if (islower(c)) {
            decrypted_text += (c - shift - 'a' + 26) % 26 + 'a';
        } else if (isupper(c)) {
            decrypted_text += (c - shift - 'A' + 26) % 26 + 'A';
        } else {
            decrypted_text += c;
        }
    }
    return decrypted_text;
}

char frequency_analysis(const string& ciphertext) {
    unordered_map<char, int> frequency;
    for (char c = 'a'; c <= 'z'; ++c) {
        frequency[c] = 0;
    }

    for (char c : ciphertext) {
        if (isalpha(c)) {
            frequency[tolower(c)]++;
        }
    }

    return max_element(frequency.begin(), frequency.end(),
                       [](const pair<char, int>& a, const pair<char, int>& b) {
                           return a.second < b.second;
                       })->first;
}

pair<string, int> cryptanalysis_attack(const string& ciphertext) {
    char most_common_in_english = 'e';
    char most_common_in_ciphertext = frequency_analysis(ciphertext);

    int shift = (most_common_in_ciphertext - most_common_in_english + 26) % 26;

    string decrypted_text = caesar_decrypt(ciphertext, shift);
    return make_pair(decrypted_text, shift);
}

int main() {
    string ciphertext = "this is pratik";
    auto result = cryptanalysis_attack(ciphertext);
    cout << "Decrypted Text: " << result.first << endl;
    cout << "Detected Shift: " << result.second << endl;
    return 0;
}

File: cns/exp3.py
import string

def caesar_decrypt(ciphertext, shift):
    decrypted_text = ""
    for char in ciphertext:
        if char in string.ascii_lowercase:
            decrypted_text += chr((ord(char) - shift - 97) % 26 + 97)
        elif char in string.ascii_uppercase:
            decrypted_text += chr((ord(char) - shift - 65) % 26 + 65)
        else:
            decrypted_text += char
    return decrypted_text

def frequency_analysis(ciphertext):
    frequency = {}
    for char in string.ascii_lowercase:
        frequency[char] = 0

    for char in ciphertext.lower():
        if char in string.ascii_lowercase:
            frequency[char] += 1

    most_common_letter = max(frequency, key=frequency.get)
    return most_common_letter

def cryptanalysis_attack(ciphertext):
    most_common_in_english = 'e'
    most_common_in_ciphertext = frequency_analysis(ciphertext)
    
    shift = (ord(most_common_in_ciphertext) - ord(most_common_in_english)) % 26

    decrypted_text = caesar_decrypt(ciphertext, shift)
    return decrypted_text, shift

# Example usage
ciphertext = "Gur dhvpx oebja sbk whzcf bire gur ynml qbt."
decrypted_text, shift = cryptanalysis_attack(ciphertext)
print(f"Decrypted Text: {decrypted_text}")
print(f"Detected Shift: {shift}")
